### ISSUE TITLE
Generic transformations of SSSOM mappings

### ISSUE THREAD

**Issue by gouttegd** [gouttegd/sssom-java#1]
The following is a rough idea for some kind of _SSSOM-to-OWL transformation language_, which would allow to specify how to generate arbitrary OWL axioms from SSSOM mappings.

By “arbitrary”, I mean in particular that the generated axioms could completely differ from the “default” [OWL axioms described in the SSSOM spec](https://mapping-commons.github.io/sssom/spec/#rdfxml-serialised-re-ified-owl-axioms), to fit any purpose beyond merely serialising a mapping set into OWL.

The primary use case would be the production of cross-species bridges for UBERON, but the point of the transformation language is that it would be completely generic, so that it could be used for the generation of any other kind of bridges.

A transformation would be specified as a pair `selector action`, where `selector` selects the mappings to which the `action` will be applied.

## Selectors
The simplest selector is `*`, it simply selects all mappings.

Other selectors are of the form `field=value`, where `field` is a SSSOM field. For example, `predicate_id=skos:exactMatch` will select all mappings with a predicate `skos:exactMatch`. For convenience, for identifier fields the `_id` suffix may be omitted, so one could use `predicate=skos:exactMatch` for the same effect.

A `*` wildcard may be used in the `value`. For example, `subject=UBERON:*` would select all mappings where the subject is a Uberon class; `subject=UBERON:6*` would further select all mappings where the subject is a Uberon class of the 6xxxxxx series. When a wildcard is present, it MUST be the last character of the value; that is, it is not possible to use `semapv:crossSpecies*Match`.

A selector can select on more than one field. For example, `object=FBbt:* mapping_justification=semapv:ManualCuration` would select all manually curated mappings where the object is a FBbt class.

Syntactic sugar: Selectors could be nested as follows:

```
predicate=skos:exactMatch {
    object=FBbt:* ACTION_FOR_FLYBASE_MAPPING;
    object=WBbt:* ACTION_FOR_WORMBASE_MAPPING;
}
```

The above would be equivalent to:

```
predicate=skos:exactMatch object=FBbt:* ACTION_FOR_FLYBASE_MAPPING;
predicate=skos:exactMatch object=WBbt:* ACTION_FOR_WORMBASE_MAPPING;
```

## Actions
An ”action” would be either a single function call (`action()`) or a block containing several function calls (`{ action1(); action2(); }`).

For now and at least in the context of the ROBOT plugin, the main function available would be a function to generate OWL axioms (hereafter named `generate`). For example, the following action would generate a cross-reference annotation on the subject pointing to the object:

```
generate(owl:AnnotationAssertion($subject, oboInOwl:hasDbXref, $object))
```

`$subject` and `$object` would be built-in variables always accessible in the context of an action, and would refer to, well, the subject and the object of the mapping being transformed, respectively.

## Example
The following full example would generate the kind of cross-species bridges we use in Uberon:

```
subject=UBERON:* predicate=semapv:crossSpeciesExactMatch object=FBbt:* {
    # generate the equivalence axiom ($object EquivalentTo $subject and (part_of some NCBITaxon:7227)
    generate(owl:EquivalentClasses($object, owl:ObjectIntersectionOf($subject, owl:ObjectSomeValuesOf(BFO:0000050, NCBITaxon:7227))));

    # generate the "OBO Foundry unique label" (IAO:0000589) on the taxon-specific term
    # by appending " (Drosophila)" to its original label
    # (the `X->Y` syntax gets the annotation Y on object X; `concat` is a built-in string concatenation function)
    generate(owl:AnnotationAssertion($object, IAO:0000589, concat($object->rdfs:label, " (Drosophila)")));
}
```

**Comment by gouttegd** [comment]
## Convenience functions

Because the action to generate the cross-species bridges is quite complex, we don’t want to repeat it for every taxon for which we want such bridges. So, there would be a possibility to declare _convenience functions_:

```
define EQUIVALENT_FOR_TAXON(taxon, name) {
    generate(owl:EquivalentClasses($object, owl:ObjectIntersectionOf($subject, owl:ObjectSomeValuesOf(BFO:0000050, $taxon))));
    generate(owl:AnnotationAssertion($object, IAO:0000589, concat($object->rdfs:label, " (", $name, ")")));
}
```

Cross-species bridges could then be specified as follows:

```
# Select all mappings with the `skos:crossSpeciesExactMatch` predicate
predicate=skos:crossSpeciesExactMatch {
    # Then select on the object for each taxon-specific ontology, and call the above function
    object=FBbt:* EQUIVALENT_FOR_TAXON(NCBITaxon:7227, "Drosophila");
    object=WBbt:* EQUIVALENT_FOR_TAXON(NCBITaxon:????, "C. Elegans"); # yeah, I am too lazy to look up the taxon ID for C. elegans! :D 
    object=ZFA:*  EQUIVALENT_FOR_TAXON(NCBITaxon:????, "Zebrafish"); # Ditto for zebrafish! :D 
    ...
}
```

**Comment by gouttegd** [comment]
## Putting everything together

With all the features described above, a single file could describe everything that Uberon needs to do for cross-species mappings:

```
# Convenience function to create a "equiv-for-taxon" bridge
define EQUIVALENT_FOR_TAXON(taxon, name) {
    generate(owl:EquivalentClasses($object, owl:ObjectIntersectionOf($subject, owl:ObjectSomeValuesFrom(obo:BFO_0000050, $taxon))));
    generate(owl:AnnotationAssertion($object, obo:IAO_0000589, concat($object->rdfs:label, " (", name, ")")));
}

# Make sure the Uberon class is always the subject for all subsequent transformations
object=UBERON:* invert();

# Inject a OBO-style cross-reference representing the mapping (for backwards compatibility)
* generate(owl:AnnotationAssertion($subject, oboInOwl:hasDbXref, $subject->id));

# Then generate the cross-species bridging axioms proper
predicate=semapv:crossSpeciesExactMatch {
    object=FBbt:* EQUIVALENT_FOR_TAXON(NCBITaxon:7227, "Drosophila");
    object=WBbt:* EQUIVALENT_FOR_TAXON(NCBITaxon:????, "C. elegans");
    object=ZFA:*  EQUIVALENT_FOR_TAXON(NCBITaxon:????, "Zebrafish");
    ...
}
```

**Comment by cmungall** [comment]
I like it! But we should be careful about unleashing Yet Another Template Language. At first I thought, why not just use robot templates? But I see the use case here also involves filters which I believe is outside what robot templates and dosdps can do. I am not sure about OTTR though.

[linkml-owl](https://github.com/linkml/linkml-owl/) provides a generic way to go from instance data to functional-syntax centric TBoxes, RBoxes and ABoxes. It provides a way of annotating a schema such as the sssom schema to do [isomorphic transformations](https://linkml.io/linkml-owl/basics/) - but this is only suited for simple transformations. For anything more complex [you can use fstrings or jinja templates](https://linkml.io/linkml-owl/templates/)

So some of the use cases here can do managed by code like

```
{% if startswith(subject_id, "UBERON:") %}
    {% if startswith(object_id, "ZFA:") %}
      EquivalentClasses({{subject_id}}, ObjectIntersectionOf({{object_id}}, ObjectSomeValuesFrom(BFO:0000050 NCBITaxon:7955))
    {% else %}
     ...
    {% endif %}
{% else %}
{% endif %}
```

This is a bit wordy, but you could imagine having [jinja macros](https://jinja.palletsprojects.com/en/3.1.x/templates/#macros) for common SSSOM patterns, and passing in tables of data from the header that perform the same function as the uberon taxon xref headers.

**Comment by gouttegd** [comment]
> But we should be careful about unleashing Yet Another Template Language.

Yeah, it wouldn’t be the first time I try to re-invent the wheel… ^^'

Thanks for the pointer to LinkML-OWL. I’ll play with that a bit.

On a cursory look, and if I understand correctly, I don’t like how the OWL template is tied to the LinkML schema. That would mean that anyone who want to have their own “OWL translation” of a SSSOM mapping set would have to fork the SSSOM schema and annotate it with their own template.

**Comment by gouttegd** [comment]
I mention for completeness an alternative proposition that was made during the “UBERON/RO/CL/PATO tech board” meeting today: basically the same idea of describing arbitrary transformations for SSSOM mappings, but using a YAML- or JSON-based syntax instead of the ad-hoc syntax above.

So for example, instead of a transformation rule described like this:

```
subject==FBbt:* predicate==semapv:crossSpeciesExactMatch (object==UBERON:* || object==CL:*) -> generate('%subject EquivalentTo %object and (BFO:0000050 some NCBITaxon:7227)');
```

it would rather look like this:

```yaml
rules:
  - filter:
      - field: subject
        value: "FBbt:*"
      - field: predicate
        value: "semapv:crossSpeciesExactMatch"
      - or_filter:
          - field: object
            value: "UBERON:*"
          - field: object
            value: "CL:*"
    actions:
      - generate: "%subject EquivalentTo %object and (BFO:0000050 some NCBITaxon:7227)"
```

I admit this could make _parsing_ the rules easier (by re-using any of the 118 YAML/JSON parsers available in modern languages), but I feel this also makes _reading_ them (by a human reader, not a machine) much more cumbersome.

**Comment by gouttegd** [comment]
This is now done.

In the end, the implementation differs slightly from the initial design. The most important difference is that the possibility of creating custom “convenience functions” is not available, since I don’t think it is worth the hassle (at least for now; I may change my mind in the future).

Here is an example of a ruleset for the FBbt-to-Uberon/CL mappings:

```
prefix UBERON:    <http://purl.obolibrary.org/obo/UBERON_>
prefix CL:        <http://purl.obolibrary.org/obo/CL_>
prefix BFO:       <http://purl.obolibrary.org/obo/BFO_>
prefix FBbt:      <http://purl.obolibrary.org/obo/FBbt_>
prefix IAO:       <http://purl.obolibrary.org/obo/IAO_>
prefix NCBITaxon: <http://purl.obolibrary.org/obo/NCBITaxon_>
prefix OIO:       <http://www.geneontology.org/formats/oboInOwl#>

[fbbt] subject==FBbt:* {
    (object==UBERON:* || object==CL:*) {
        predicate==semapv:crossSpeciesExactMatch -> gen('%subject_id EquivalentTo %object_id and (BFO:0000050 some NCBITaxon:7227)');
        predicate==semapv:crossSpeciesCloseMatch -> gen('%subject_id SubclassOf %object_id');
        predicate==* -> gen('%subject_id Annotation IAO:0000589 "%subject_label (Drosophila)"');
    }
}

[xrefs] predicate==* -> gen('%subject_id Annotation OIO:hasDbXref "%object_curie"');
```

Full documentation of the final syntax will arrive soon.
